### PR TITLE
ATP v3 update

### PR DESCRIPTION
**Pull request by jaredoconnell**
## Changes introduced with this PR

This update adds the changes that correspond to the go SDK ATP v3 update: https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/57

The error reporting in this version is **massively** improved. Now errors in the ATP and the plugin are properly reported to the ATP client, which is then reported by the engine.

Some of the suggestions by Webb were also addressed in this branch.

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by jdowni000 — APPROVED** [review]
lgtm

**Review by jdowni000 — APPROVED** [review]

**Comment by jdowni000** [comment]
Fix that build error real quick should be good to go

**Review by jdowni000 — APPROVED** [review]
Nice!

**Review by mfleader — APPROVED** [review]

File: src/arcaflow_plugin_sdk/atp.py [hunk 1]
```diff
@@ -247,49 +319,53 @@ def read_hello(self) -> HelloMessage:
...
+            run_id,
+        {
                 "signal_id": signal_id,
                 "data": input_data,
             }
         )
 
     def send_client_done(self):
-        self.send_runtime_message(MessageType.CLIENT_DONE, {})
+        self.send_runtime_message(MessageType.CLIENT_DONE, "", "")
```
Comment (webbnh) [inline src/arcaflow_plugin_sdk/atp.py:349]: Granted, the type of the third parameter of `send_runtime_message` is `any`, but all the previous calls have sent `dict`'s...it seems like settling on a particular type would be good.  (If you really want to send nothing and don't like `{}`, how about `None`?)

File: src/arcaflow_plugin_sdk/atp.py [hunk 2]
```diff
@@ -247,49 +319,53 @@ def read_hello(self) -> HelloMessage:
...
+                "id": message_type,
+                "run_id": run_id,
                 "data": data,
             }
         )
-        self.stdin.flush()
+        self.to_server_pipe.flush()
 
-    def read_results(self) -> (str, any, str):
+    def read_single_result(self) -> (str, str, any, str):
```
Comment (webbnh) [inline src/arcaflow_plugin_sdk/atp.py:361]: Tuples are problematic from a maintenance perspective.  Consider using a [`namedtuple`](https://docs.python.org/3/library/collections.html#namedtuple-factory-function-for-tuples-with-named-fields) (or [`NamedTuple`](https://docs.python.org/3.6/library/typing.html?highlight=namedtuple#typing.NamedTuple)) or defining a [`@dataclass`](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass) instead, which allow you to name the members, which frees you from having to get the ordering right and allows you to add context to the values.

**Review by webbnh — COMMENTED** [review]
Hi @jaredoconnell, I didn't make it all the way through, but here are some thoughts for you.

File: src/arcaflow_plugin_sdk/atp.py [hunk 3]
```diff
@@ -64,24 +70,26 @@ class HelloMessage:
...
-            stdout: io.FileIO,
+            input_pipe: io.FileIO,
+            output_pipe: io.FileIO,
             stderr: io.FileIO,
     ) -> None:
-        self.stdin = stdin
-        self.stdout = stdout
+        self.input_pipe = input_pipe
+        self.output_pipe = output_pipe
         self.stderr = stderr
```
[the remark this answers is not in the document]
Comment (jaredoconnell) [reply src/arcaflow_plugin_sdk/atp.py:93]: Will be resolved in the next PR. I learned abut the differences between instance and class attributes after I wrote this part, but I forgot to go back and fix it.

File: src/arcaflow_plugin_sdk/atp.py [hunk 4]
```diff
@@ -91,121 +99,185 @@ def run_plugin(
...
+                    try:
+                        self.handle_signal(run_id, signal_msg)
+                    except Exception as e:
+                        self.send_error_message(run_id, False, False,
+                                                f"Exception while handling signal: {e} {traceback.format_exc()}")
+                elif msg_id == MessageType.CLIENT_DONE:
                     return
                 else:
+                    self.send_error_message(run_id, False, False, f"Unknown runtime message ID: {msg_id}")
                     self.stderr.write(f"Unknown kind of runtime message: {msg_id}")
```
[the remark this answers is not in the document]
Comment (jaredoconnell) [reply src/arcaflow_plugin_sdk/atp.py:171]: Good catch. It should also use the new send_error_message format.
I'll fix that, and add the keyword parameters in the next PR.

**Pull request merged**